### PR TITLE
Forward pointers

### PR DESCRIPTION
**Pull request by kjames88**
Creates but does not use forward pointers so far.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **c689fbf**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #637 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/nrodvVaO14QZSOAL1xmUFTTHZUxwWVD0)

**Comment by no2chem** [comment]
This is a great start, but in order to be performant, the forward pointers need to be added during a read rather than via a special call. I would make it an optional parameter to the read call.

**Comment by kjames88** [comment]
Doesn't this have an issue with caching?  I can wait until the backpointers are all recovered and then write out a forward list, but if the cache is large enough for performance, haven't we then already completed all the reads in question?

**Comment by no2chem** [comment]
@kjames88 I mean the second. It should be a significant performance benefit, because there is no extra call to teach the logunit the new forward pointer, by adding a second call you pay the cost of two message roundtrips.

So a flow would be with entries A (1), B (2) C (3):
1) get backpointer from sequencer to C @ 3
2) read C @ 3 which contains a backpointer to B @ 2
3) read B @ 2 with a hint that we came from C @ 3, which will populate the forward pointer B->3, if not present.
4) read A @ 1 with a hint that we came from B @ 2, which will populate the forward pointer for A->2, if not present.

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 1]
```diff
@@ -223,6 +221,16 @@ private void trim(CorfuPayloadMsg<TrimRequest> msg, ChannelHandlerContext ctx, I
...
     }
 
+    @ServerHandler(type = CorfuMsgType.POINTER_UPDATE)
+    private void setPointer(CorfuPayloadMsg<PointerUpdate> msg, ChannelHandlerContext ctx, IServerRouter r,
+                            boolean isMetricsEnabled) {
+        PointerUpdate payload = msg.getPayload();
+        log.debug("setPointer {}->{}", payload.getCurrent(), payload.getNext());
+        forwardPointerMap.putIfAbsent(payload.getStream(), new HashMap<>());
+        forwardPointerMap.get(payload.getStream()).put(payload.getCurrent(), payload.getNext());
+        //r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
```
Comment (dahliamalkhi) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: I guess we should remove this comment

File: runtime/src/main/java/org/corfudb/runtime/view/replication/IReplicationProtocol.java [hunk 2]
```diff
@@ -35,6 +35,17 @@
...
+    /** Write forward pointer to the log for a given address
+     *
+     * This function is non-blocking
+     *
+     * @param layout
+     * @param streamId
+     * @param address
+     * @param next
+     */
+    void writeForwardpointer(Layout layout, UUID streamId, long address, long next); /* throws in case of mismatched pointer? */
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/replication/IReplicationProtocol.java]: again, should we delete the comment already?

File: runtime/src/main/java/org/corfudb/runtime/view/replication/QuorumReplicationProtocol.java [hunk 3]
```diff
@@ -131,6 +132,10 @@ public void write(Layout layout, ILogData data) throws OverwriteException {
         }
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public void writeForwardpointer(Layout layout, UUID stream, long address, long nextAddress) {
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/replication/QuorumReplicationProtocol.java]: any thoughts on this? should we care that only a quorum might have the actual entry?

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java (hunk 1 above)
Comment (kjames88) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: Or uncomment it.  Should we ack or just do async?

File: runtime/src/main/java/org/corfudb/runtime/view/replication/IReplicationProtocol.java (hunk 2 above)
Comment (kjames88) [inline runtime/src/main/java/org/corfudb/runtime/view/replication/IReplicationProtocol.java]: It raises a question:  do we need to catch an error here or assume the pointers always match?

File: runtime/src/main/java/org/corfudb/runtime/view/replication/QuorumReplicationProtocol.java (hunk 3 above)
Comment (kjames88) [inline runtime/src/main/java/org/corfudb/runtime/view/replication/QuorumReplicationProtocol.java]: Quorum raises the question of how much coordination is needed to get the pointers replicated (are we just making copies or running paxos like the data replication)?

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0466ba5**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #637 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/bmC5buHtGoFbr5LtVqYxO918yE0Zf1XJ)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **a9b4e71**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #637 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/bsBrRoJKc5uMtYYYnPdvWGycs0AJntPK)

**Comment by dahliamalkhi** [comment]
I am going to close this PR, but keep the branch.

closing status summary: there are two serious issues with the current PR. One is implementation: we want to piggyback forward pointer messages on reads, in order to avoid adding an extra round-trip.

The second one is fundamental, hence closing this: Log entries may contain contents for multiple stream. However,  for every log entry, only the first stream-read actually fetches data from the logging-unit. Remaining stream-reads on the same entries will hit the read cache. Hence, only the first stream-read will provide forward-pointers.

**Pull request closed without merging**